### DISCUSSION TITLE
请问有没有什么方法能够查询某一天有数据的所有设备

### DISCUSSION THREAD

**Discussion by herowzz** [apache/iotdb#3929] [Q&A] [answered]
假如我有几万个设备，想查询某一天有数据的设备（比如查询2020-01-01这一天有数据的所有设备）
我尝试用 `select last ... where time >= 2020-01-01 ` 查询，但是这样的话，比如某个设备从2020-03-01开始有数据，但是2020-01-01没有数据，这个设备也会被查询出来， 而且last查询只能用大于号，小于号并不支持，所以没法限定范围
想请教一下有没有什么简单的办法？

**Comment by SteveYurongSu** [comment, marked as answer]
目前可以通过下面的步骤实现：

1. 先查询指定时间范围所有（涉及的）设备下的 count 数量，例如
```sql
select count(measurement) from root.storage_group.*.*.* where time>2020-01-01T00:00:00 and time < 2020-01-02T00:00:00
```

这个过程中，您可能还会用到**路径层级分组聚合**，可参考：https://iotdb.apache.org/zh/UserGuide/Master/IoTDB-SQL-Language/DML-Data-Manipulation-Language.html#%E8%B7%AF%E5%BE%84%E5%B1%82%E7%BA%A7%E5%88%86%E7%BB%84%E8%81%9A%E5%90%88

2. 取 count 不为 0 的设备（序列） 名称作为您后续的查询输入

---

另外，您在查询过程中可能会遇到

![image](https://user-images.githubusercontent.com/30497621/132834895-91aeb5ba-5f76-41be-b498-dbdebe409528.png)

此时，您只需要改一下 `iotdb-engine.propeties` 配置文件里面的 `max_deduplicated_path_num` 然后执行 `LOAD CONFIGURATION `或者重启系统即可。
